### DISCUSSION TITLE
[RFC] Tags and scopes (enforce module boundaries)

### DISCUSSION THREAD

**Discussion by haukurmar** [vercel/turborepo#7460] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

When you are building a mono-repo you want to avoid having circular dependencies between packages, so you would like to define what packages can import other packages.

NX has a good way of doing this with what they call module boundaries using tags and scopes define in the project.json, then the linter takes over.

https://nx.dev/core-features/enforce-module-boundaries

an example of this would look something like this.

app: web
package: ui
package: config
package: auth

The app "web" imports on all of these packages 
The package "ui" should not be able to import from any of these since it's stand-alone
The package "auth" can reference "config", but not ui or web.

I hope this explains it a bit, otherwise you can look into how NX is doing it.

### Describe the solution you'd like

I am not sure how this should be implemented, perhaps an inspiration from NX and how they solve it.

It would be nice that this could be configured in a turbo.json inside each package and/or on the root of the workspace.

### Describe alternatives you've considered

Alternative is to use NX as it provides this feature and a lot more.

**Comment by NicholasLYang** [comment]
Hey @haukurmar, thanks for the suggestion! I'll add it to our list of potential features. We're definitely making ergonomics and observability a top priority, and this feature seems to fall under those categories.

**Comment by austinbiggs** [comment]
I'd like to echo the benefits of module boundaries. I'm pretty eager to migrate my Nx monorepo to Turborepo, but losing the guardrails that help to prevent circular dependencies is a turn off.

**Reply by anthonyshew** [reply]
Minor note: Turborepo always errors on circular dependencies so that specific concern is taken care of today. 👍

(There's certainly still a ton of value to be added by a true module boundary enforcement feature - but just to speak to that concern in particular!)

**Reply by haukurmar** [reply]
Can you elaborate a bit more? 
I am using Turborepo in one of my projects and I know of many circular dependencies between packages and I see no errors.

**Reply by kaoz70** [reply]
I have been using Turborepo (with Yarn classic or pnpm) and I have not seen Turborepo error on this, could you explain where this happens? Maybe I am missing something.

**Reply by donnellyjoe** [reply]
I was looking at `nx` recently and got jealous of the tags/lint feature. I then found [this ESLint plugin](https://github.com/javierbrea/eslint-plugin-boundaries#readme) which has similar functionality. I don't love how it is configured by pattern matching against paths to packages though. Adding/removing/renaming packages would require manually updating the lint configuration. I would love to see something like this implemented directly within turbo, but I'm not exactly what the ergonomics of it would look like

**Reply by anthonyshew** [reply]
@haukurmar @kaoz70 When Turborepo finds circular dependencies in the package graph, it should be erroring. If you have examples of this not happening, could you open a GitHub Issue with a reproduction?

**Reply by mikaelbalin** [reply]
There are definitely some errors and warnings in case of circular dependencies.

<img width="1115" alt="Screenshot 2024-11-12 at 11 52 53" src="https://github.com/user-attachments/assets/0560698f-4faf-46a9-9343-12843693cfae">

**Comment by anthonyshew** [comment]
Hey, all! Let's fold this discussion and interest into https://github.com/vercel/turborepo/discussions/9435. Thank you!

**Reply by shinebayar-g** [reply]
And this discussion is _closed_?

**Reply by anthonyshew** [reply]
Sorry! Meant here: https://github.com/vercel/turborepo/discussions/9435

Edited my message above.
